File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        l , r = 0 , 0
        hashmap = {}
        result = 0
        for r in range(len(fruits)):
            if fruits[r] in hashmap:
                hashmap[fruits[r]] += 1
            else:
                hashmap[fruits[r]] = 1
            while len(hashmap) > 2:
                hashmap[fruits[l]] -= 1
                if hashmap[fruits[l]] == 0:
                    hashmap.pop(fruits[l])
                l += 1
            result = max(result, r - l + 1)
        return result
            

        

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.c
int totalFruit(int* fruits, int fruitsSize) {
    int result = 0;
    int l = 0,r,t1=-1,t2=-1;
    for(r = 0; r < fruitsSize; r++){

        if(t1 == -1)
            t1 = fruits[r];
        else if(t2 == -1 && fruits[r] != t1)
            t2 = fruits[r];
        else{
            if(t1 != fruits[r] && t2 != fruits[r]){
                l = r - 1;
                while(l > 0 && fruits[l] == fruits[l-1]) l--;
                    t1 = fruits[l];
                    t2 = fruits[r];
            }
        }
        result = (result > r - l + 1) ? result : r - l + 1;
        



    }
    return result;
    
}